src/client/java/daos-java/src/main/native/io_daos_dfs_DaosFsClient.c

```diff
@@ -824,6 +874,79 @@ Java_io_daos_dfs_DaosFsClient_dfsRead(JNIEnv *env, jobject client,
 	return size;
 }
 
+static inline void
+decode_dfs_desc(char *buf, dfs_desc_t **desc_ret, daos_event_t **event_ret,
+	uint64_t *offset_ret, uint64_t *len)
+{
+    uint64_t dfs_mem;
+    uint16_t eid;
+    dfs_desc_t *desc;
+
+    memcpy(&dfs_mem, buf, 8);
+    desc = (dfs_desc_t *)dfs_mem;
+    *desc_ret = desc;
+    desc->sgl.sg_nr = 1;
+    desc->sgl.sg_nr_out = 0;
+    desc->size = 0;
+    buf += 24; /* skip native handle, data mem address and eq handle */
+    memcpy(offset_ret, buf, 8);
+    buf += 8;
+    memcpy(len, buf, 8);
+    buf += 8;
+    desc->iov.iov_len = desc->iov.iov_buf_len = (size_t)(*len);
+    /* event */
+    memcpy(&eid, buf, 2);
+    *event_ret = desc->eq->events[eid];
+}
+
+static int
+update_actual_size(void *udata, daos_event_t *ev, int ret)
+{
+	dfs_desc_t *desc = (dfs_desc_t *)udata;
+	char *desc_buffer = desc->ret_buf_address;
+	uint32_t value = (uint32_t)desc->size;
+
+	memcpy(desc_buffer, &ret, 4);
+	desc_buffer += 4;
+	memcpy(desc_buffer, &value, 4);
+	ev->ev_error = 0;
+}
+
+JNIEXPORT void JNICALL
+Java_io_daos_dfs_DaosFsClient_dfsReadAsync(JNIEnv *env, jobject client,
+	jlong dfsPtr, jlong objId, jlong descBufAddress)
+{
+    dfs_t *dfs = *(dfs_t **)&dfsPtr;
+    dfs_obj_t *file = *(dfs_obj_t **)&objId;
+    char *buf = (char *)descBufAddress;
+    uint64_t offset;
+    uint64_t len;
+    dfs_desc_t *desc;
+    daos_event_t *event;
+    int rc;
+
+    decode_dfs_desc(buf, &desc, &event, &offset, &len);
+    rc = daos_event_register_comp_cb(event,
+    			update_actual_size, desc);
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs